x-pack/packages/observability/observability_ai/observability_ai_server/root_cause_analysis/run_root_cause_analysis.ts

```diff
@@ -0,0 +1,305 @@
+downstream dependencies, and all related entities have been investigated without
+discovering new anomalies, it may be appropriate to end the process.
+  - If all investigated entities (e.g., services, hosts, containers) are
+functioning normally, with no relevant issues found, and there are no signs of
+dependencies being affected, you may consider ending the process.
+  - Avoid concluding the investigation based solely on symptoms or the absence
+of immediate errors in the data. Unless a system change has been connected to
+the incident, it is important to continue investigating dependencies to ensure
+the root cause has been accurately identified.`;
+
+export function runRootCauseAnalysis({
+  serviceName,
+  start: requestedStart,
+  end: requestedEnd,
+  esClient,
+  alertsClient,
+  rulesClient,
+  observabilityAIAssistantClient,
+  spaceId,
+  indices,
+  connectorId,
+  inferenceClient,
+  context: initialContext,
+  logger: incomingLogger,
+  prevEvents,
+}: {
+  context: string;
+  serviceName: string;
+  logger: Logger;
+  inferenceClient: InferenceClient;
+  start: number;
+  end: number;
+  alertsClient: AlertsClient;
+  rulesClient: RulesClient;
+  esClient: ObservabilityElasticsearchClient;
+  observabilityAIAssistantClient: ObservabilityAIAssistantClient;
+  indices: {
+    logs: string[];
+    traces: string[];
+    sloSummaries: string[];
+  };
+  connectorId: string;
+  spaceId: string;
+  prevEvents?: RootCauseAnalysisEvent[];
+}): Observable<RootCauseAnalysisEvent> {
+  const logger = incomingLogger.get('rca');
+
+  const entity = { 'service.name': serviceName };
+
+  const bucketSize = calculateAuto
+    .atLeast(30, moment.duration(requestedEnd - requestedStart))!
+    .asMilliseconds();
+
+  const start = Math.floor(requestedStart / bucketSize) * bucketSize;
+  const end = Math.floor(requestedEnd / bucketSize) * bucketSize;
+
+  const initialMessage = {
+    role: MessageRole.User as const,
+    content: `Investigate the health status of ${formatEntity(entity)}.
+    
+    The context given for this investigation is:
+
+    ${initialContext}`,
+  };
+
+  const nextEvents = [initialMessage, ...(prevEvents ?? [])];
+
+  const initialRcaContext: RootCauseAnalysisContext = {
+    connectorId,
+    start,
+    end,
+    esClient,
+    events: nextEvents,
+    indices,
+    inferenceClient,
+    initialContext,
+    alertsClient,
+    observabilityAIAssistantClient,
+    logger,
+    rulesClient,
+    spaceId,
+    tokenLimit: 32_000,
+  };
+
+  const investigationTimeRangePrompt = `## Time range
+  
+    The time range of the investigation is ${new Date(start).toISOString()} until ${new Date(
+    end
+  ).toISOString()}`;
+
+  initialContext = `${initialContext}
+
+  ${investigationTimeRangePrompt}
+  `;
+
+  const next$ = callTools(
+    {
+      system: RCA_SYSTEM_PROMPT_BASE,
+      connectorId,
+      inferenceClient,
+      messages: nextEvents,
+      logger,
+    },
```

Comment:
Shouldn't this also include a `tools` property? Or how are the tools specified?

Reply:
... I'm likely missing something since this would be easy to catch with Typescript.